internal/util/vault_sa.go

```diff
@@ -0,0 +1,246 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed setting up token for %s/%s: %w", kms.Tenant, kms.tenantSAName, err)
+	}
+
+	err = kms.initCertificates(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize Vault certificates: %w", err)
+	}
+	// connect to the Vault service
+	err = kms.connectVault()
+	if err != nil {
+		return nil, err
+	}
+
+	return kms, nil
+}
+
+// Destroy removes the temporary stored token from the ServiceAccount and
+// destroys the vaultTenantConnection object.
+func (kms *VaultTenantSA) Destroy() {
+	if kms.saTokenDir != "" {
+		_ = os.RemoveAll(kms.saTokenDir)
+	}
+
+	kms.vaultTenantConnection.Destroy()
+}
+
+// setServiceAccountName stores the name of the ServiceAccount in the
+// configuration if it has been set in the options.
+func (kms *VaultTenantSA) setServiceAccountName(config map[string]interface{}) error {
+	err := setConfigString(&kms.tenantSAName, config, "tenantSAName")
+	if errors.Is(err, errConfigOptionInvalid) {
+		return err
+	}
+
+	return nil
+}
+
+// getServiceAccount returns the Tenants ServiceAccount with the name
+// configured in the VaultTenantSA.
+func (kms *VaultTenantSA) getServiceAccount() (*corev1.ServiceAccount, error) {
+	c := kms.getK8sClient()
+	sa, err := c.CoreV1().ServiceAccounts(kms.Tenant).Get(context.TODO(),
+		kms.tenantSAName, metav1.GetOptions{})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ServiceAccount %s/%s: %w", kms.Tenant, kms.tenantSAName, err)
+	}
+
+	return sa, nil
+}
+
+// getToken looks up the ServiceAccount and the Secrets linked from it. When it
+// finds the Secret that contains the `token` field, the contents is read and
+// returned.
+func (kms *VaultTenantSA) getToken() (string, error) {
+	sa, err := kms.getServiceAccount()
+	if err != nil {
+		return "", err
+	}
+
+	c := kms.getK8sClient()
+	for _, secretRef := range sa.Secrets {
+		secret, err := c.CoreV1().Secrets(kms.Tenant).Get(context.TODO(), secretRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", fmt.Errorf("failed to get Secret %s/%s: %w", kms.Tenant, secretRef.Name, err)
+		}
+
+		token, ok := secret.Data["token"]
+		if ok {
+			return string(token), nil
+		}
+	}
+
+	return "", fmt.Errorf("failed to find token in ServiceAccount %s/%s", kms.Tenant, kms.tenantSAName)
+}
+
+// getTokenPath creates a temporary directory structure that contains the token
+// linked from the ServiceAccount. This path can then be used in place of the
+// standard `/var/run/secrets/kubernetes.io/serviceaccount/token` location.
+func (kms *VaultTenantSA) getTokenPath() (string, error) {
+	dir, err := ioutil.TempDir("", kms.tenantSAName)
+	if err != nil {
+		return "", fmt.Errorf("failed to create directory for ServiceAccount %s/%s: %w", kms.tenantSAName, kms.Tenant, err)
+	}
+
+	token, err := kms.getToken()
+	if err != nil {
+		return "", err
+	}
+
+	err = ioutil.WriteFile(dir+"/token", []byte(token), 0600)
```

Comment:
define a new variable for `dir+/token`? as it is used in multiple places?

Reply:
Only used 2x, in the same function. I do not think it is needed or making things clearer when defining a constant for it.

Reply:
yes currently used only in 2 places. if something changes in the future same changes need to be made in multiple places